O24 dataset

PhillCli opened the pull request:
## Description
adds O24 dataset [jcp paper](https://doi.org/10.1063/5.0043793) with two variants:
* `O24by5mb.py` - geometries with ghosted H as a midbond functions located at COM
* `O24by5.py` - geometries without midbond functions

## Questions
- [ ] Should I modify any install script to accommodate additional files?
- [ ] Should test files be added, e.g. for three subsets (ED, DD, MX) that are defined?
- [x] I've defined constant `KCALMOL2WAVENUMBERS = 349.7551` inside the database, as paper values are in cm^-1 and psi database driver expects kcal/mol, is there a cleaner way to handle this conversion?

## Checklist
- [ ] Tests added for any new features
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge

JonathonMisiewicz commented:
I can leave the other questions to people more familiar with the database structure, but for your third question, `from psi4.driver import constants` and then `constants.kcalmol2wavenumbers`

PhillCli commented:
updated the PR:
* removed unused imports
* switched to psi4.drivers.constants for conversion factors
* reformatted files with yapf, I had malfunctioning commit hook when submitting original PR.

loriab commented on psi4/share/psi4/databases/O24by5.py:
```diff
@@ -0,0 +1,2628 @@
+        2,
+        [
+            1,
+        ],
```
formatting like this is the only reason I prefer black to lgtm

PhillCli replied:
tweaked a bit to make it more readable

loriab commented on psi4/share/psi4/databases/O24by5.py:
```diff
@@ -0,0 +1,2628 @@
+        [
+            2,
+        ],
+    )
```
is there a reason to not allow the non-CP case for these databases, like https://github.com/psi4/psi4/blob/master/psi4/share/psi4/databases/S22.py#L879-L880  and in ACTV?

PhillCli replied:
enabled it, but not sure how it works (if at all) with `mb` variant.

loriab reviewed (commented):
thanks for the new databases!

no need to add to an install, as they're globbed here https://github.com/psi4/psi4/blob/master/psi4/CMakeLists.txt#L328

you're welcome to add tests, but not every database file is represented in tests. the docs build will ensure that it can import.

PeterKraus commented on psi4/share/psi4/databases/O24by5mb.py:
```diff
@@ -0,0 +1,2858 @@
+from psi4.driver import constants
+
+# <<< O24by5 Database Module >>>
+dbse = "O24by5mb"
```
Is there no way to combine the two variants into a single file? The midbond positions could easily be added into the geometry database automatically (as we always seem to add `Gh(H) @ 0.0 0.0 0.0`), the only thing I'm not sure the runner knows how to do is to select a `GEOM` based on the `dbse` that's used to call this...

loriab replied:
maybe have the non-mb database file define the geometry strings in a dict, then strip the last two lines before feeding into Molecule constructor in GEOS (units, etc. can be first lines in the string), then the mb database imports the plain one, deep copies all the data structures and uses the full strings in GEOS.

Or, database files are pretty static, and these two are easy to diff to check sync. I can live with the duplication

PeterKraus replied:
If it's fine by you, it's fine by me. I was just wondering if we were missing a trick there, but if it'd have to be hacked to avoid duplication, those few LoC are worth a more obvious solution.

loriab commented on psi4/share/psi4/databases/O24by5.py:
```diff
@@ -0,0 +1,2628 @@
+units angstrom
+no_reorient
+no_com
+symmetry c1
```
yeah, I'd drop the `no_reorient`, `no_com`, `symmetry c1` if they weren't added with deliberation. Spelling out the 0.9 (R_eq vs. Ang) in the docstring as Peter suggests would also be nice. otherwise, good to go, imo.

PhillCli replied:
Since these files were originally used for some of the results in the paper, I wanted them to reflect reference calculations with other software (including symmetry used).
Purged, everything except for `angstrom` as it should work without them and might save some resource, where spacial symmetry is applicable.

PhillCli commented:
hello, I'll try to work on the replies and changes to have it completed by the end of the week, might I ask what's the timeframe we're working here before 1.4 stable release?

loriab commented:
> hello, I'll try to work on the replies and changes to have it completed by the end of the week, might I ask what's the timeframe we're working here before 1.4 stable release?

We're probably looking at rc3 asap and stable 1.4 next week. So end-of-the-week is fine.

As I see it, the items are:
* remove no_com, etc. if not really needed
* clarify in docstrings whether 1.9 is R_eq or AA
* allow uncp in `ACTV` rather than hardcoding both it and `ACTV_CP` to cp, unless there's a good reason not to
* the suggestion about deduplicating the two files might be possible, but I don't think it's worth it for a db
* I don't recall why CI/Azure was failing on your last commit, but there were certainly points where it was psi4 master's fault, so you may need to rebase.

loriab commented:
Hi, did you get a chance to look at this further? So long as as big problems aren't uncovered in last week's rc3, we'd like to do the 1.4 release including the new databases this week. I think the items in https://github.com/psi4/psi4/pull/2154#issuecomment-870728272 were pretty small. Let me know if you need advice for the rebase.

PhillCli commented:
> Hi, did you get a chance to look at this further? So long as as big problems aren't uncovered in last week's rc3, we'd like to do the 1.4 release including the new databases this week. I think the items in [#2154 (comment)](https://github.com/psi4/psi4/pull/2154#issuecomment-870728272) were pretty small. Let me know if you need advice for the rebase.

> remove no_com, etc. if not really
-  [x] done 

> clarify in docstrings whether 1.9 is R_eq or AA
-  [x] done, I think. See the latest changes `R=0.9R_eq`.

> allow uncp in ACTV rather than hardcoding both it and ACTV_CP to cp, unless there's a good reason not to
-  [x] done, TBH I've worked exclusively with mid-bond version with CP, I don't know if unCP with mid-bonds makes even sense. Need someone with `unCP` experience to comment on that.

> I don't recall why CI/Azure was failing on your last commit, but there were certainly points where it was psi4 master's fault, so you may need to rebase.
-  [x] rebased

as for DRYifing two files, I don't think it's desired to couple these two files, as they should be in-principle, runnable standalone.

PhillCli commented on psi4/share/psi4/databases/O24by5.py:
```diff
@@ -0,0 +1,2628 @@
+
+# <<< Comment Lines >>>
+TAGL = {}
+TAGL["%s-%s" % (dbse, "1-0.9")] = """ CN He R=0.9"""
```
clarified, that this is wrt to `R_eq`, see the changes. Thank you for the feedback!

loriab reviewed (approved):
lgtm, thanks for the databases!

JonathonMisiewicz commented on psi4/share/psi4/databases/O24by5.py:
```diff
@@ -0,0 +1,2264 @@
+    BIND[key] = value / constants.kcalmol2wavenumbers
+
+# <<< Comment Lines >>>
+TAGL = {}
```
These are very repetitive dictionary-filling statements. Can I convince you to rewrite this as a set of for loops? One over `DIST` and one over `enumerate(new_array_containing_dimer_names, start=1)`. The same goes in the other file.

PhillCli replied:
I'll try to set it up, shortly.

JonathonMisiewicz reviewed (approved):
One comment to shorten the files, but LGTM otherwise.

PeterKraus reviewed (commented):
I'm fine with the PR as is. If you tidy up the BIND and TAGL dictionaries as Jonathon suggested above, even better.

PhillCli commented:
> I'm fine with the PR as is. If you tidy up the BIND and TAGL dictionaries as Jonathon suggested above, even better.

did the part with `TAGL`, not sure about `BIND` since assigned values should be unique per each entry. I would prefer to keep it in a declarative fashion, instead of cooking up some scary bind values array accessed by index.

JonathonMisiewicz commented:
Agreed - that's why I recommended TAGL re-structuring but never mentioned BIND. Thanks for this - PR is now good to go.

PhillCli commented:
@loriab Since, I've touched this code once more. I tried to add a test if all types (`mb`,`non-mb`) x (`cp=1, `cp=0`) passes (for a selected subset)

I've run into for `midbond` (three fragments) with `cp=0`:
```
python3 tests/python/databases/input.py


felez: [array([1.]), array([1., 0.])]
all_fc_known: True
all_fm_known: True
zel: 2.0
fzel: [1.0, 1.0]
c: [0.0, 0]
fc: [0.0]
fc: [0.0]
m: [1]
fm: [1]
fm: [1]
Assess candidate (0.0, (0.0, 0.0), 1, (1, 1)): 1   2   3   4   4-0 4-1 5   5-0 5-1 6   6-0 6-1 7   7-0 7-1
Assess candidate (0.0, (0.0, 0.0), 1, (1, 1)): T   T   T   T   T   T   T           T   T   T   T   T   T   --> False
Following warning is harmless if you've altered chgmult through `set_molecular_change` or `set_multiplicity`. Such alterations are an expert feature. Specifying in the original molecule string is preferred.


felez: [array([1.]), array([1., 0.])]
all_fc_known: False
all_fm_known: False
zel: 2.0
fzel: [1.0, 1.0]
c: [0.0, 0]
fc: [0.0, 0.0]
fc: [0.0, 0.0]
m: [1]
fm: [1, 1, 2]
fm: [1, 1, 2]
Assess candidate (0.0, (0.0, 0.0), 1, (1, 1)): 1   2   3   4   4-0 4-1 5   5-0 5-1 6   7   8  
Assess candidate (0.0, (0.0, 0.0), 1, (1, 1)): T   T   T   T   T   T   T           T   T   T   --> False
Assess candidate (0.0, (0.0, 0.0), 1, (1, 2)): T   T   T   T   T   T   T       T   T   T       --> False
Assess candidate (0.0, (0.0, 0.0), 1, (2, 1)): T   T   T   T   T   T   T   T       T   T       --> False
Assess candidate (0.0, (0.0, 0.0), 1, (2, 2)): T   T   T   T   T   T   T   T   T   T   T       --> False
Traceback (most recent call last):
  File "/home/filip/software/psi4-1.4rc-opt/lib/psi4/driver/qcdb/molecule.py", line 1468, in to_dict
    validated_molrec = qcel.molparse.from_arrays(speclabel=False, verbose=0, domain='qm', **molrec)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/from_arrays.py", line 381, in from_arrays
    verbose=verbose)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/chgmult.py", line 485, in validate_and_fill_chgmult
    c_final, fc_final, m_final, fm_final = reconcile(cgmp_exact_c, cgmp_exact_fc, cgmp_exact_m, cgmp_exact_fm)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/chgmult.py", line 475, in reconcile
    raise ValidationError(err)
qcelemental.exceptions.ValidationError: Inconsistent or unspecified chg/mult: sys chg: 0.0, frag chg: [0.0, 0.0], sys mult: 1, frag mult: [1, 1]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/python/databases/input.py", line 10, in <module>
    psi4.wrapper_database.database('scf', 'O24by5mb', cp=0, subset=["4-0.9", "12-0.9", "20-0.9", "23-0.9"])
  File "/home/filip/software/psi4-1.4rc-opt/lib/psi4/driver/wrapper_database.py", line 452, in database
    molecule = core.Molecule.from_dict(GEOS[rgt].to_dict())
  File "/home/filip/software/psi4-1.4rc-opt/lib/psi4/driver/qcdb/molecule.py", line 1477, in to_dict
    validated_molrec = qcel.molparse.from_arrays(speclabel=False, verbose=0, domain='qm', **molrec)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/from_arrays.py", line 381, in from_arrays
    verbose=verbose)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/chgmult.py", line 485, in validate_and_fill_chgmult
    c_final, fc_final, m_final, fm_final = reconcile(cgmp_exact_c, cgmp_exact_fc, cgmp_exact_m, cgmp_exact_fm)
  File "/home/filip/software/psi4-1.4rc-opt/lib/qcelemental/molparse/chgmult.py", line 475, in reconcile
    raise ValidationError(err)
qcelemental.exceptions.ValidationError: Inconsistent or unspecified chg/mult: sys chg: 0.0, frag chg: [None, None], sys mult: 1, frag mult: [None, None]
```

I'm inclined to discourage using it :(

PhillCli commented:
ayy caramba, already merged before my final push :(

JonathonMisiewicz commented:
Apologies. When I merged this, I didn't see any signals that there was anything left to do for this PR.

Others can correct me if I'm wrong, but I think that at this point, any further changes would need to be in a separate PR.

loriab commented:
Agree that it's good to discourage mb=True + np=False. Since you've already got them ready, please make another PR. If you delete the branch name on GH, you can open new PR with same branch name w/o rebasing in master, if you like.

PhillCli commented:
follow up in #2252
